src/canisters/frontend/ic-asset/src/batch_upload/plumbing.rs

```diff
@@ -46,30 +49,69 @@ pub(crate) struct ProjectAsset {
     pub(crate) encodings: HashMap<String, ProjectAssetEncoding>,
 }
 
+type IdMapping = BTreeMap<usize, Nat>;
+type UploadQueue = Vec<(usize, Vec<u8>)>;
 pub(crate) struct ChunkUploader<'agent> {
     canister: Canister<'agent>,
     batch_id: Nat,
+    api_version: u16,
     chunks: Arc<AtomicUsize>,
     bytes: Arc<AtomicUsize>,
+    // maps uploader_chunk_id to canister_chunk_id
+    id_mapping: Arc<Mutex<IdMapping>>,
+    upload_queue: Arc<Mutex<UploadQueue>>,
 }
+
 impl<'agent> ChunkUploader<'agent> {
-    pub(crate) fn new(canister: Canister<'agent>, batch_id: Nat) -> Self {
+    pub(crate) fn new(canister: Canister<'agent>, api_version: u16, batch_id: Nat) -> Self {
         Self {
             canister,
             batch_id,
+            api_version,
             chunks: Arc::new(AtomicUsize::new(0)),
             bytes: Arc::new(AtomicUsize::new(0)),
+            id_mapping: Arc::new(Mutex::new(BTreeMap::new())),
+            upload_queue: Arc::new(Mutex::new(vec![])),
         }
     }
 
+    /// Returns an uploader_chunk_id, which is different from the chunk id on the asset canister.
+    /// uploader_chunk_id can be mapped to canister_chunk_id using `uploader_ids_to_canister_chunk_ids`
+    /// once `finalize_upload` has completed.
     pub(crate) async fn create_chunk(
         &self,
         contents: &[u8],
         semaphores: &Semaphores,
-    ) -> Result<Nat, CreateChunkError> {
-        self.chunks.fetch_add(1, Ordering::SeqCst);
+    ) -> Result<usize, CreateChunkError> {
+        let uploader_chunk_id = self.chunks.fetch_add(1, Ordering::SeqCst);
         self.bytes.fetch_add(contents.len(), Ordering::SeqCst);
-        create_chunk(&self.canister, &self.batch_id, contents, semaphores).await
+        if contents.len() == MAX_CHUNK_SIZE || self.api_version < 2 {
+            let canister_chunk_id =
+                create_chunk(&self.canister, &self.batch_id, contents, semaphores).await?;
+            let mut map = self.id_mapping.lock().await;
+            map.insert(uploader_chunk_id, canister_chunk_id);
+            println!("directly uploaded {uploader_chunk_id}");
```

Comment:
intentionally left in? If so, this should probably be `info!`

Reply:
That's an oversight, thanks!